sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -434,6 +434,27 @@ object SimplifyConditionals extends Rule[LogicalPlan] with PredicateHelper {
     case _ => false
   }
 
+  /**
+   * Condition for redundant null check based on intolerant expressions.
+   * @param ifNullExpr expression that takes place if checkedExpr is null
+   * @param ifNotNullExpr expression that takes place if checkedExpr is not null
+   * @param checkedExpr expression that is checked for null value
+   */
+  private def isRedundantNullCheck(
+      ifNullExpr: Expression,
+      ifNotNullExpr: Expression,
+      checkedExpr: Expression): Boolean = {
+    val isNullIntolerant = ifNotNullExpr.find { x =>
+      !x.isInstanceOf[NullIntolerant] && x.find(e => e.semanticEquals(checkedExpr)).nonEmpty
```

Comment:
The same logic? https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala#L105-L109

Reply:
Actually i think we need slightly different logic. Consider these two examples where `x` will be the null-checked column: 

1. `substring(x, coalesce(a, b), c)`

2. `substring(coalesce(x, d), a, c)`

For 1. we need to be null-intolerant (even though `coalesce` is null-tolerant), so if `x` is null, we replace the `substring` with null value no matter what are the other children. For 2. we need to be null-tolerant and we will not replace the `substring` by null value. So we need to check the expression with respect to the position of `x` (the column that is being null-checked). Does it make sense?

Reply:
Probably, you meant `FiterExec.isNullIntolerant(ifNotNullExpr) || additional checks for the case having null-tolerant exprs inside ifNotNullExpr`? (`FiterExec.isNullIntolerant` is private though...)

Reply:
Yeah, `FilterExec.isNullIntolerant(ifNotNullExpr)` is a stronger condition than we need so in case there is null-tolerant expr inside we need to check if the null-checked column is in its subtree. Using the logic from `FilterExec.isNullIntolerant` the function could look like this:
```
def isNullIntolerant(expr: Expression): Boolean = expr match {
  case e: NullIntolerant => e.children.forall(isNullIntolerant)
  case e if e.find(x => x.semanticEquals(checkedExpr)).isEmpty => true
  case _ => false
}
```

Reply:
Ah, I see. For better code readability, could you split the condition into the two parts as I suggested above? Also, I think its better to leave some comments about why we need more checks there.

Reply:
I agree that the committed code is not very intuitive so i can think of this way which seems to be more readable (added also some comments):

```
private def isRedundantNullCheck(
    ifNullExpr: Expression,
    ifNotNullExpr: Expression,
    checkedExpr: Expression): Boolean = {
  
  // checks if expr is null-intolerant with respect to checkedExpr
  def isNullIntolerant(expr: Expression): Boolean = expr match {
    case e: NullIntolerant => e.children.forall(isNullIntolerant)
    // if some child is null-tolerant but the checkedEpxr is not in its subtree
    // we can still consider the whole expr as null-intolerant 
    // with respect to checkedExpr
    case e if e.find(x => x.semanticEquals(checkedExpr)).isEmpty => true
    case _ => false
  }
  
  isNullIntolerant(ifNotNullExpr) && {
    (ifNullExpr.semanticEquals(checkedExpr) ||
      (ifNullExpr.foldable && ifNullExpr.eval() == null)) &&
      // we still need to make sure that checkedExpr is inside ifNotNullExpr
      ifNotNullExpr.find(x => x.semanticEquals(checkedExpr)).nonEmpty
  }
}
```
But not sure if this is what you had in mind when suggesting to split the condition. Can you think of a better way how to compose this?

Reply:
Hmmm, that still looks complicated.. If we cannot avoid the complexity for the stronger condition, as another option, I think we can cover the simple case (`FiterExec.isNullIntolerant(ifNotNullExpr)`) only in this pr. If necessary, we might be able to optimize the condition in future work. I think keeping the code simple is more important. WDYT?

Reply:
Well, the think is that if we use the simple version with `FilterExec.isNullIntolerant(ifNutNullExpr)` we will loose (because of the recursive check) all expressions that contain literals (because literals are null-tolerant), so for example expressions like this `substring(title#5, 0, 3)` will not be included in the optimization (which the jira was targeted for in the first place). So I suggest one of these 2 options:

1. Use the complex version of the code and thus include more expressions in the optimization
2. Have the code more simple and use the original version before the generalization step, i.e. 
```
private def isRedundantNullCheck(
    ifNullExpr: Expression,
    ifNotNullExpr: Expression,
    checkedExpr: Expression): Boolean = {
  ifNotNullExpr.isInstanceOf[NullIntolerant] && {
    (ifNullExpr == checkedExpr || ifNullExpr == Literal.create(null, checkedExpr.dataType)) &&
      ifNotNullExpr.children.contains(checkedExpr)
  }
}
```
where `checkedExpr` must be direct child and thus we don't have to check the whole subtree for null-intolerance (so expressions that have Literals in the subtree are still included).
I am fine with either of these 2 options. What do you think?